lightning/src/ln/channelmanager.rs

```diff
@@ -72,6 +72,258 @@ use core::ops::Deref;
 #[cfg(any(test, feature = "std"))]
 use std::time::Instant;
 
+mod inbound_payment {
+	use bitcoin::hashes::{Hash, HashEngine};
+	use bitcoin::hashes::cmp::fixed_time_eq;
+	use bitcoin::hashes::hmac::{Hmac, HmacEngine};
+	use bitcoin::hashes::sha256::Hash as Sha256;
+	use chain::keysinterface::{KeyMaterial, KeysInterface, Sign};
+	use ln::{PaymentHash, PaymentPreimage, PaymentSecret};
+	use ln::msgs;
+	use ln::msgs::MAX_VALUE_MSAT;
+	use util::chacha20::ChaCha20;
+	use util::logger::Logger;
+
+	use core::convert::TryInto;
+	use core::ops::Deref;
+
+	const IV_LEN: usize = 16;
+	const METADATA_LEN: usize = 16;
+	const METADATA_KEY_LEN: usize = 32;
+	const AMT_MSAT_LEN: usize = 8;
+	// Used to shift the payment type bits to take up the top 3 bits of the metadata bytes, or to
+	// retrieve said payment type bits.
+	const METHOD_TYPE_OFFSET: usize = 5;
+
+	/// A set of keys that were HKDF-expanded from an initial call to
+	/// [`KeysInterface::get_inbound_payment_key_material`].
+	///
+	/// [`KeysInterface::get_inbound_payment_key_material`]: crate::chain::keysinterface::KeysInterface::get_inbound_payment_key_material
+	pub(super) struct ExpandedKey {
+		/// The key used to encrypt the bytes containing the payment metadata (i.e. the amount and
+		/// expiry, included for payment verification on decryption).
+		metadata_key: [u8; 32],
+		/// The key used to authenticate an LDK-provided payment hash and metadata as previously
+		/// registered with LDK.
+		ldk_pmt_hash_key: [u8; 32],
+		/// The key used to authenticate a user-provided payment hash and metadata as previously
+		/// registered with LDK.
+		user_pmt_hash_key: [u8; 32],
+	}
+
+	impl ExpandedKey {
+		pub(super) fn new(key_material: &KeyMaterial) -> ExpandedKey {
+			hkdf_extract_expand(&vec![0], &key_material)
```

Comment:
If you have a sec before Jeff reviews, could you change the salt here to some string? eg `b"LDK Inbound Payment Key Expansion"`?